### PR TITLE
[EX-20] Setup local signin on backend and frontend

### PR DESCRIPTION
**Pull request by tomaszferens**
image

image

image

### CODE REVIEW

File: backend/src/auth/auth.service.ts [hunk 1]
```diff
@@ -1,20 +1,35 @@
...
+    constructor(
+        private readonly usersService: UsersService,
+        private readonly jwtService: JwtService,
+    ) {}
 
-    async create(createUserDto: CreateUserDto): Promise<User> {
-        const alreadyCreatedUser = await this.userModel.findOne({
-            email: createUserDto.email,
-        })
+    async validateUser(email: string, password: string): Promise<any> {
```
Comment (marckraw) [inline backend/src/auth/auth.service.ts]: return types ^^

File: backend/src/auth/jwt.strategy.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+export class JwtStrategy extends PassportStrategy(Strategy) {
+    constructor(private readonly usersService: UsersService) {
+        super({
+            jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+            ignoreExpiration: false,
+            secretOrKey: jwtConstants.secret,
+        })
+    }
+
+    async validate(payload: TokenData) {
```
Comment (marckraw) [inline backend/src/auth/jwt.strategy.ts]: should this have return type ? I think we should decide on some rules about where we insert types and where not

File: backend/src/auth/local.strategy.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    constructor(
+        private readonly authService: AuthService,
+        private readonly usersService: UsersService,
+    ) {
+        super({
+            usernameField: 'email',
+        })
+    }
+
+    async validate(email: string, password: string): Promise<any> {
```
Comment (marckraw) [inline backend/src/auth/local.strategy.ts]: return type

File: backend/src/models/TokenData.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export interface TokenData {
+    sub: string
```
Comment (marckraw) [inline backend/src/models/TokenData.ts]: what `sub` means ? maybe i don't have context thats why i dont understand that :D

File: backend/src/users/users.service.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import { Model } from 'mongoose'
+import { omit } from 'lodash'
+
+import { User, userFieldsToOmit } from '../models/user'
+
+@Injectable()
+export class UsersService {
+    constructor(@InjectModel('User') private readonly userModel: Model<User>) {}
+
+    async create(user: Omit<User, '_id'>): Promise<Model<User>> {
```
Comment (marckraw) [inline backend/src/users/users.service.ts]: just a question, this `Omit<User, '_id'>` is a Typescript way to omit some type from interface ?

File: frontend/src/Signin.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                        type={status.loading ? 'button' : 'submit'}
+                        fullWidth
+                        variant="contained"
+                        color="primary"
+                        className={classes.submit}
+                    >
+                        {status.loading ? (
+                            <CircularProgress size={24} color="inherit" />
+                        ) : (
+                            props.t('Sign in')
```
Comment (marckraw) [inline frontend/src/Signin.js]: lets desctructure translation aswell

File: frontend/src/Signup.js [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+                    Sign up
+                </Typography>
+                <Form className={classes.form}>
+                    <Grid container spacing={2}>
+                        <Grid item xs={12} sm={6}>
+                            <TextField
+                                name="firstName"
+                                variant="outlined"
+                                fullWidth
+                                label={props.t('First name')}
```
Comment (marckraw) [inline frontend/src/Signup.js]: desctructure translations maybe

**Comment by marckraw** [comment]
Also, did you add endpoints toi Postman collection ? :)
image

File: backend/src/models/TokenData.ts (hunk 4 above)
Comment (tomaszferens) [reply backend/src/models/TokenData.ts]: According to some documentation:

JWT RFC 7519 defines some standard claim names in “4.1. Registered Claim Names”.

`sub - Subject`

```
The "sub" (subject) claim identifies the principal that is the
   subject of the JWT.  The claims in a JWT are normally statements
   about the subject.  The subject value MUST either be scoped to be
   locally unique in the context of the issuer or be globally unique.
   The processing of this claim is generally application specific.  The
   "sub" value is a case-sensitive string containing a StringOrURI
   value.  Use of this claim is OPTIONAL.
```
In our case it's user's id

File: backend/src/auth/local.strategy.ts (hunk 3 above)
Comment (tomaszferens) [reply backend/src/auth/local.strategy.ts]: will get rid of that. TS is smart enough to infer the types based on what you actually returned

File: backend/src/auth/auth.service.ts (hunk 1 above)
Comment (tomaszferens) [reply backend/src/auth/auth.service.ts]: as below

File: backend/src/auth/jwt.strategy.ts (hunk 2 above)
Comment (tomaszferens) [reply backend/src/auth/jwt.strategy.ts]: as below

File: backend/src/users/users.service.ts (hunk 5 above)
Comment (tomaszferens) [reply backend/src/users/users.service.ts]: yes, it's a built in interface

**Comment by tomaszferens** [comment]
> Also, did you add endpoints toi Postman collection ? :)
> image

Yes, dodaned.

File: backend/src/models/TokenData.ts (hunk 4 above)
Comment (marckraw) [reply backend/src/models/TokenData.ts]: ah ok, so just some kind of convention name, cool ;)

**Comment by marckraw** [comment]
> > Also, did you add endpoints toi Postman collection ? :)
> > image
> 
> Yes, dodaned.

a to nie jest rejestracja ? a to co tutaj zrobiles to tez logowanie ? ;p

**Pull request merged**